ambry-router/src/test/java/com.github.ambry.router/GetOperationTest.java

```diff
@@ -0,0 +1,474 @@
+            operationCallback, time);
+    ArrayList<RequestInfo> requestListToFill = new ArrayList<>();
+    requestRegistrationCallback.requestListToFill = requestListToFill;
+
+    for (MockServer server : mockServerLayout.getMockServers()) {
+      server.setServerErrorForAllRequests(ServerErrorCode.Blob_Not_Found);
+    }
+
+    while (!op.isOperationComplete()) {
+      op.poll(requestRegistrationCallback);
+      List<ResponseInfo> responses = sendAndWaitForResponses(requestListToFill);
+      for (ResponseInfo response : responses) {
+        op.handleResponse(response);
+        if (op.isOperationComplete()) {
+          break;
+        }
+      }
+    }
+
+    Assert.assertEquals("Must have attempted sending requests to all replicas", replicasCount,
+        correlationIdToGetOperation.size());
+    Assert.assertTrue("Operation should be complete at this time", op.isOperationComplete());
+    Assert.assertTrue(operationFuture.isDone());
+    RouterException routerException = (RouterException) operationException.get();
+    Assert.assertEquals(RouterErrorCode.BlobDoesNotExist, routerException.getErrorCode());
+  }
+
+  /**
+   * Test the case with Blob_Not_Found errors from most servers, and Blob_Deleted at just one server. The latter
+   * should be the exception received for the operation.
+   * @throws Exception
+   */
+  @Test
+  public void testErrorPrecedenceWithBlobDeletedCase()
+      throws Exception {
+    GetBlobInfoOperation op =
+        new GetBlobInfoOperation(routerConfig, mockClusterMap, responseHandler, blobIdStr, operationFuture,
+            operationCallback, time);
+    ArrayList<RequestInfo> requestListToFill = new ArrayList<>();
+    requestRegistrationCallback.requestListToFill = requestListToFill;
+
+    int i = 0;
+    int indexToSetBlobDeleted = random.nextInt(replicasCount);
+    for (MockServer server : mockServerLayout.getMockServers()) {
+      if (i == indexToSetBlobDeleted) {
+        server.setServerErrorForAllRequests(ServerErrorCode.Blob_Deleted);
+      } else {
+        server.setServerErrorForAllRequests(ServerErrorCode.Blob_Not_Found);
+      }
+      i++;
+    }
+
+    while (!op.isOperationComplete()) {
+      op.poll(requestRegistrationCallback);
+      List<ResponseInfo> responses = sendAndWaitForResponses(requestListToFill);
+      for (ResponseInfo response : responses) {
+        op.handleResponse(response);
+        if (op.isOperationComplete()) {
+          break;
+        }
+      }
+    }
+
+    Assert.assertTrue("Operation should be complete at this time", op.isOperationComplete());
+    Assert.assertTrue(operationFuture.isDone());
+    RouterException routerException = (RouterException) operationException.get();
+    Assert.assertEquals(RouterErrorCode.BlobDeleted, routerException.getErrorCode());
+  }
+
+  /**
+   * Test the case with Blob_Not_Found errors from most servers, and Blob_Expired at just one server. The latter
```

Comment:
That sounds like a good idea. It can make the tests more sophisticated though and needs some thinking through. For now, I have made a small change based on your suggestion, in the way deleted and expired cases are tested, and perhaps that method can evolve to incorporate what you suggest.